receiver/snmpreceiver/config.go

```diff
@@ -0,0 +1,592 @@
+	if cfg.PrivacyType == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyPrivacyType)
+	} else {
+		switch strings.ToUpper(cfg.PrivacyType) {
+		case "DES", "AES", "AES192", "AES192C", "AES256", "AES256C": // ok
+		default:
+			combinedErr = multierr.Append(combinedErr, errBadPrivacyType)
+		}
+	}
+
+	return combinedErr
+}
+
+// validateMetricConfigs validates all MetricConfigs, AttributeConfigs, and ResourceAttributeConfigs
+func validateMetricConfigs(cfg *Config) error {
+	var combinedErr error
+
+	// Validate the Attribute and ResourceAttribute configs up front
+	combinedErr = multierr.Append(combinedErr, validateAttributeConfigs(cfg))
+	combinedErr = multierr.Append(combinedErr, validateResourceAttributeConfigs(cfg))
+
+	// Ensure there is at least one MetricConfig
+	metrics := cfg.Metrics
+	if len(metrics) == 0 {
+		return multierr.Append(combinedErr, errMetricRequired)
+	}
+
+	// Make sure each MetricConfig has valid info
+	for metricName, metricCfg := range metrics {
+		if metricCfg.Unit == "" {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoUnit, metricName))
+		}
+
+		if metricCfg.Gauge == nil && metricCfg.Sum == nil {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoGaugeOrSum, metricName))
+		}
+
+		if len(metricCfg.ScalarOIDs) == 0 && len(metricCfg.ColumnOIDs) == 0 {
+			combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgMetricNoOIDs, metricName))
+		}
+
+		if metricCfg.Gauge != nil {
+			combinedErr = multierr.Append(combinedErr, validateGauge(metricName, metricCfg.Gauge))
+		}
+
+		if metricCfg.Sum != nil {
+			combinedErr = multierr.Append(combinedErr, validateSum(metricName, metricCfg.Sum))
+		}
+
+		if len(metricCfg.ScalarOIDs) > 0 {
+			for _, scalarOID := range metricCfg.ScalarOIDs {
+				combinedErr = multierr.Append(combinedErr, validateScalarOID(metricName, scalarOID, cfg))
+			}
+		}
+
+		if len(metricCfg.ColumnOIDs) > 0 {
+			for _, columnOID := range metricCfg.ColumnOIDs {
+				combinedErr = multierr.Append(combinedErr, validateColumnOID(metricName, columnOID, cfg))
+			}
+		}
+	}
+
+	return combinedErr
+}
+
+// validateColumnOID validates a ColumnOID
+func validateColumnOID(metricName string, columnOID ColumnOID, cfg *Config) error {
+	var combinedErr error
+
+	// Ensure that it contains an OID
+	if columnOID.OID == "" {
+		combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgColumnOIDNoOID, metricName))
+	}
+
+	// Keep track of whether the different indexed values can be differentiated by either attribute within the same metric
+	// or by different resource attributes (in different resources)
+	hasIndexedIdentifier := false
+
+	// Check that any Attributes have a valid Name and a valid Value (if applicable)
+	if len(columnOID.Attributes) > 0 {
+		for _, attribute := range columnOID.Attributes {
+			if attribute.Name == "" {
+				combinedErr = multierr.Append(combinedErr, fmt.Errorf(errMsgColumnAttributeNoName, metricName))
```

Comment:
All these nested ifs are a bit hard to read. At any point there's an `if-else` where the `if` is the end of the logic block we can just return `combinedErr` that way there doesn't need to be so much nesting in `else` blocks.

Reply:
yeah, i usually hate stuff like this. glad you pointed it out. i refactored to have less nesting